### PR TITLE
HPCC-32478 Add check for signed ECL macros

### PR DESCRIPTION
**Pull request by jackdelv**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
A few general comments (I don't necessarily know the answers)

This looks like a potential backdoor.  E.g. if a macro is signed, can you pass in a magical parameter that allows you access to something you shouldn't, otherwise.

What happens about macros called from macros.  Which macro needs to be signed?  Both, inner, outer?  Need to be very clear about the semantics, with some clear examples to demonstrate.

Associating the signing certificate with the restricted operation.  Often the engines will check if the operation is valid - by looking at the signedBy attribute on an activity.    (See calls to getGpgSignature() and similar in the parser.)  This will not associate the signer, so the restricted operations may also fail.

Main comment is this needs very careful definition of the semantics, especially on weird edge case, and plenty of test cases to demonstrate the expected behaviour.

**Comment by jackdelv** [comment]
@ghalliday I agree this isn't correct. I am currently investigating the issue.

**Comment by jackdelv** [comment]
@ghalliday I changed the checkSignedMacros method to check for an implicitly signed IFileContents object in the lexer. I also propogate the value of isImplicitlySigned if inSignedModule is false. I have included a set of examples that pass and fail as expected.

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -9766,9 +9766,22 @@ void HqlGram::checkDerivedCompatible(IIdAtom * name, IHqlExpression * scope, IHq
     }
 }
 
+bool HqlGram::checkSignedMacros()
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:9769]: style: (picky).  The prefix "check" implies that a function is going to look at something and then perform an action based on the result.  Better to have a name that fits well inside a boolean context e.g. if ([is]insideSignedMacro())

File: testing/regress/ecl/signingPass.ecl [hunk 2]
```diff
@@ -0,0 +1,49 @@
+IMPORT STD;
```
Comment (ghalliday) [inline testing/regress/ecl/signingPass.ecl:1]: This needs a matching testing/regress/ecl/key/signingPass.xml

File: testing/regress/ecl/signingFail.ecl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+//version version=3
+//version version=4
+
+import  ^ as root;
+v := #IFDEFINED(root.version, 1);
+
+IMPORT Std;
+
+#IF (v = 0)
+    Std.Outer_Macro1US(4);
```
Comment (ghalliday) [inline testing/regress/ecl/signingFail.ecl:14]: These macros are not defined, and should not be in Std.  I suspect it will require a signed file checked into this directory.

File: ecl/hql/hqlgram2.cpp [hunk 4]
```diff
@@ -9766,9 +9766,22 @@ void HqlGram::checkDerivedCompatible(IIdAtom * name, IHqlExpression * scope, IHq
...
+{
+    if (inSignedModule)
+        return true;
+
+    IFileContents * contents = lexObject->query_FileContents();
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:9774]: Better to encapsulate the logic within the lexObject.  i.e.
```
lexObject->isImplicitilySigned()

bool HqlLex::isImplicitlySigned()
{
    return contents && contents->isImplicitlySigned();
}
```
However I am not sure that function will correctly handle all macros.  e.g.if an attributes calls an unsigned macro which uses a signed macro within it.

I would expect the following implementation:

```
bool HqlLex::isImplicitlySigned()
{
    if (contents && contents->isImplicitlySigned())
        return true;
    if (!inmacro)
        return false;
    return inmacro->isImplicitlySigned();
}
```

**Review by ghalliday — COMMENTED** [review]
@jackdelv a few comments, but looks clean.
The tests need to have the macros checked in - so it is possible to see what is being tested.

File: ecl/hql/hqllex.l [hunk 5]
```diff
@@ -1542,7 +1542,7 @@ FUNCTIONMACRO|MACRO {
                         Owned<IFileContents> macroContents;
                         if (!isFunctionMacro)
                         {
-                            macroContents.setown(createFileContentsFromText(len, lexer->yyBuffer+startpos, lexer->sourcePath, lexer->yyParser->inSignedModule, lexer->yyParser->gpgSignature, 0));
+                            macroContents.setown(createFileContentsFromText(len, lexer->yyBuffer+startpos, lexer->sourcePath, lexer->yyParser->inSignedModule /* Propogate isImplicitlySigned to child IFileContents */ || (lexer->query_FileContents() && lexer->query_FileContents()->isImplicitlySigned()), lexer->yyParser->gpgSignature, 0));
```
[the remark this answers is not in the document]
Comment (jackdelv) [reply ecl/hql/hqllex.l]: Fixed.

File: testing/regress/ecl/signingPass.ecl (hunk 2 above)
Comment (jackdelv) [reply testing/regress/ecl/signingPass.ecl]: Added key file.

File: testing/regress/ecl/signingFail.ecl (hunk 3 above)
Comment (jackdelv) [reply testing/regress/ecl/signingFail.ecl]: Added macros to testing/regress/ecl/signing

**Review by ghalliday — COMMENTED** [review]
@jackdelv looks good, and a really good set of test cases.
I can't yet think of a way to force that option on (the system deliberately stops it being overridden to prevent a security hole).

One comment.  Otherwise please turn into a real PR, squash and target master (too dangerous to go in a point release, and a minor release should be coming soon.)

File: ecl/hql/hqllex.l [hunk 6]
```diff
@@ -1542,7 +1542,9 @@ FUNCTIONMACRO|MACRO {
                         Owned<IFileContents> macroContents;
                         if (!isFunctionMacro)
                         {
-                            macroContents.setown(createFileContentsFromText(len, lexer->yyBuffer+startpos, lexer->sourcePath, lexer->yyParser->inSignedModule, lexer->yyParser->gpgSignature, 0));
+                            // Propogate isImplicitlySigned to child IFileContents 
+                            bool contentsIsSigned = lexer->yyParser->inSignedModule || (lexer->query_FileContents() && lexer->query_FileContents()->isImplicitlySigned());
```
[the remark this answers is not in the document]
Comment (jackdelv) [reply ecl/hql/hqllex.l]: Yes, I have changed it use insideSignedMacro.

**Review by dcamper — APPROVED** [review]
Looks good. Bonus points for good tests!
